templates/switchers.js

```diff
@@ -1,197 +1,194 @@
+ * @private
+ */
+const _navigate_to_first_existing = (urls) => {
+  // Navigate to the first existing URL in urls.
+  for (const url of urls) {
     fetch(url)
-      .then(function(response) {
+      .then((response) => {
         if (response.ok) {
           window.location.href = url;
-        } else {
-          navigate_to_first_existing(urls);
         }
       })
-      .catch(function(error) {
-        navigate_to_first_existing(urls);
-      });
-  }
-
-  function on_version_switch() {
-    const selected_version = this.options[this.selectedIndex].value + '/';
-    const url = window.location.href;
-    const current_language = language_segment_from_url();
-    const current_version = version_segment_from_url();
-    const new_url = url.replace('/' + current_language + current_version,
-                                '/' + current_language + selected_version);
-    if (new_url != url) {
-      navigate_to_first_existing([
-        new_url,
-        url.replace('/' + current_language + current_version,
-                    '/' + selected_version),
-        '/' + current_language + selected_version,
-        '/' + selected_version,
-        '/'
-      ]);
-    }
+      .catch((err) => void err);
   }
 
-  function on_language_switch() {
-    let selected_language = this.options[this.selectedIndex].value + '/';
-    const url = window.location.href;
-    const current_language = language_segment_from_url();
-    const current_version = version_segment_from_url();
-    if (selected_language == 'en/') // Special 'default' case for English.
-      selected_language = '';
-    let new_url = url.replace('/' + current_language + current_version,
-                              '/' + selected_language + current_version);
-    if (new_url != url) {
-      navigate_to_first_existing([
-        new_url,
-        '/'
-      ]);
-    }
+  // if all else fails, redirect to the d.p.o root
+  window.location.href = '/';
+};
+
+/**
+ * Callback for the version switcher.
+ * @param {Event} event
+ * @returns {void}
+ * @private
+ */
+const _on_version_switch = (event) => {
+  if (_IS_LOCAL) return;
+
+  const selected_version = event.target.value;
+  // English has no language prefix.
+  const new_prefix_en = `/${selected_version}/`;
+  const new_prefix =
+    _CURRENT_LANGUAGE === 'en'
+      ? new_prefix_en
+      : `/${_CURRENT_LANGUAGE}/${selected_version}/`;
+  if (_CURRENT_PREFIX !== new_prefix) {
+    // Try the following pages in order:
+    // 1. The current page in the current language with the new version
+    // 2. The current page in English with the new version
+    // 3. The documentation home in the current language with the new version
+    // 4. The documentation home in English with the new version
+    _navigate_to_first_existing([
+      window.location.href.replace(_CURRENT_PREFIX, new_prefix),
+      window.location.href.replace(_CURRENT_PREFIX, new_prefix_en),
+      new_prefix,
+      new_prefix_en,
```

Comment:
I'm wondering if it would be better to stick in the current language, even if that means going back to the homepage?

It might be surprising to switch to a completely different language set; as you navigate further everything is English and not your chosen language.

And I think different languages should nearly always have the same structure and same pages for any given version? (Although with some untranslated pages.)

That would mean:


```suggestion
    // 1. The current page in the current language with the new version
    // 2. The documentation home in the current language with the new version
    // 3. The current page in English with the new version
    // 4. The documentation home in English with the new version
    _navigate_to_first_existing([
      window.location.href.replace(_CURRENT_PREFIX, new_prefix),
      new_prefix,
      window.location.href.replace(_CURRENT_PREFIX, new_prefix_en),
      new_prefix_en,
```

But I'm really not sure. Might be worth asking people who regularly uses the docs in another language?

Reply:
Shall we create an issue for this? The intent in this PR was for as few behavioural changes as possible (the ideally none!).

Although, the current order does have some logic, given that a reader probably prefers to have the information (which can be translated via browser tools, Google, etc) rather than always being sent to the homepage for a language that doesn't have much translation coverage.

A